### PR TITLE
hostnamed: stop caching so much

### PR DESCRIPTION
**Pull request by poettering**
Let's always report current settings instead of what we read while initializing. It's less confusing to users, as it makes hostnamed a much thinner layer around the backing logic.

Prompted by this thread:

 https://lists.freedesktop.org/archives/systemd-devel/2020-April/044395.html

### CODE REVIEW

File: man/sd_bus_add_object.xml [hunk 1]
```diff
@@ -465,6 +465,25 @@
           This corresponds to the <constant>org.freedesktop.systemd1.Explicit</constant> annotation
           in introspection data.</para></listitem>
         </varlistentry>
+
+        <varlistentry>
+          <term><constant>SD_BUS_VTABLE_SENSITIVE</constant></term>
+
+          <listitem><para>Mark this vtable method entry as processing sensitive data. When set incoming
```
Comment (keszybz) [inline man/sd_bus_add_object.xml]: "When set, "

File: src/hostname/hostnamed.c [hunk 2]
```diff
@@ -369,6 +428,65 @@ static int property_get_hostname(
...
+                const char *interface,
+                const char *property,
+                sd_bus_message *reply,
+                void *userdata,
+                sd_bus_error *error) {
+
+        sd_bus_slot *slot;
+        Context *c;
+
+        /* The the current context without the property userdata offset added */
```
Comment (keszybz) [inline src/hostname/hostnamed.c]: "The the". Dunno, even with the comment, it took me a moment to figure out what is going on here. Maybe extend the comment a bit?

File: src/hostname/hostnamed.c [hunk 3]
```diff
@@ -369,6 +428,65 @@ static int property_get_hostname(
...
```
Comment (keszybz) [inline src/hostname/hostnamed.c]: "The the"

**Comment by poettering** [comment]
force pushed with the suggested changes. no other changes.

(i added a pretty long and nice comment to that property handling code, i hope this is very clear now...)

**Comment by poettering** [comment]
(eventually we should add the same logic to localed and timedated, btw)

**Comment by keszybz** [comment]
```
======================================================================
FAIL: test_transient_hostname_with_static (__main__.DnsmasqClientTest)
transient hostname is not applied if static hostname exists
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/autopkgtest.bc9Q5g/build.YAG/systemd/test/networkd-test.py", line 794, in test_transient_hostname_with_static
    self.assertEqual(socket.gethostname(), "foobarqux")
AssertionError: 'testgreen' != 'foobarqux'
- testgreen
+ foobarqux


----------------------------------------------------------------------
Ran 36 tests in 217.007s

FAILED (failures=1, skipped=2)
autopkgtest [21:27:51]: test networkd-test.py: -----------------------]
networkd-test.py     FAIL non-zero exit status 1
```
Looks related ;(

**Comment by keszybz** [comment]
... it is quite possible that the error is in the test, but it needs to be reconciled anyway.

**Comment by poettering** [comment]
Split the first four commits into #15804. Hopefully they pas and we can merge them before the others.

**Comment by poettering** [comment]
So seams I finally tracked this one down. The fix was trivial ultimately: not only when clients query the hostname and other data from hostnamed we need to possibly refresh the data from what#s on disk, but also when we make changes to it, since we actually take the status quo into consideration too when figuring out how the updated data is to be propagated to the kernel's setting, and similar.

So I added a few calls to re-read data on disk in the setters in addition to the getters and everything worked. 

@keszybz already liked this once, given the trivial fix I figure it should be OK to set the green label again?

**Comment by keszybz** [comment]
LGTM.

**Comment by keszybz** [comment]
bionic-amd64 is still outstanding, but it's been slow lately, and I don't think this is in any way arch-specific, so let's not wait for it.

**Pull request merged**